simple_network_sim/inference.py

```diff
@@ -0,0 +1,708 @@
+        self.inferred_variables = inferred_variables
+
+    @staticmethod
+    def generate_from_priors(fitter: ABCSMC) -> Particle:
+        """ Generate a particle from the prior distribution. All parameters in the
+        particle are independent and therefore we simply random each parameter from
+        its prior distribution. The fitter object provides the random state used for
+        random variable generation but also the prior parameters, which are
+        the values in the data pipeline.
+
+        :return: New particle generated from prior
+        """
+        return Particle({variable_name: variable_class.generate_from_prior(fitter)
+                         for variable_name, variable_class in Particle.inferred_variables_classes.items()})
+
+    def generate_perturbated(self) -> Particle:
+        """ From current particle, add a perturbation to every parameter and return
+        a newly created perturbated particle.
+
+        :return: New particle which is similar to self up to a perturbation
+        """
+        perturbed_variables = {}
+        for name, inferred_variable in self.inferred_variables.items():
+            perturbed_variables[name] = inferred_variable.generate_perturbated()
+
+        return Particle(perturbed_variables)
+
+    def validate_particle(self) -> bool:
+        """ Checks that the particle is valid, simply checks that all parameters inside
+        respect their supports, i.e. the values are attainable. It may return False for
+        example if the uniform perturbation pushed a lognormally-distribution value
+        below 0.
+
+        :return: Whether the particle is valid
+        """
+        return all(variable.validate() for variable in self.inferred_variables.values())
+
+    @staticmethod
+    def resample_and_perturbate(
+            particles: List[Particle],
+            weights: List[float],
+            random_state: np.random.Generator
+    ) -> Particle:
+        """ Resampling part of ABC-SMC, selects randomly a new particle from the
+        list of previously accepted. Then perturb it slightly. This causes
+        the persistence and selection of fit particles in a manner similar to
+        evolution algorithms. The perturbation is done until the candidate is valid,
+        this is because the perturbation is unaware of the support of the distribution
+        of the particle.
+
+        :param particles: List of particles from previous population
+        :param weights: List of weights of particles from previous population
+        :param random_state: Random state for random sampling into particles
+        :return: Newly created particles sampled from previous population and perturbated
+        """
+        while True:
+            particle = random_state.choice(particles, p=weights / np.sum(weights))
+            particle = particle.generate_perturbated()
+
+            if particle.validate_particle():
+                return particle
```

Comment:
This is probably a good idea yes.

Reply:
Just to confirm: I've enhanced the random perturbation mecanism so that particles are never rejected anymore. This was possible because we use a uniform perturbation which we can simple cap and floor at the bounds of the distribution support (typically > 0), and a truncated uniform is still uniform. If we used gaussian perturbation, it would not be correct, as a truncated gaussian distribution isn't gaussian anymore.